### ISSUE TITLE
🐛 [Bug Report] [WoTLK] Compile Errors

### ISSUE THREAD

**Issue by adellaci** [cmangos/issues#3666]
### Bug Details

```
In file included from /usr/include/c++/9/bits/stl_algobase.h:64,
                 from /usr/include/c++/9/bits/specfun.h:45,
                 from /usr/include/c++/9/cmath:1927,
                 from /home/mangos/cmangos/wotlk/src/shared/Common.h:28,
                 from /home/mangos/cmangos/wotlk/src/framework/Dynamic/ObjectRegistry.h:22,
                 from /home/mangos/cmangos/wotlk/src/framework/Dynamic/FactoryHolder.h:24,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/UnitAI.h:23,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/CreatureAI.h:22,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/CreatureAI.cpp:19:
/usr/include/c++/9/bits/stl_pair.h: In instantiation of ‘struct std::pair<const unsigned int, LootTemplate>’:
/usr/include/c++/9/ext/aligned_buffer.h:91:28:   required from ‘struct __gnu_cxx::__aligned_buffer<std::pair<const unsigned int, LootTemplate> >’
/usr/include/c++/9/bits/hashtable_policy.h:233:43:   required from ‘struct std::__detail::_Hash_node_value_base<std::pair<const unsigned int, LootTemplate> >’
/usr/include/c++/9/bits/hashtable_policy.h:279:12:   required from ‘struct std::__detail::_Hash_node<std::pair<const unsigned int, LootTemplate>, false>’
/usr/include/c++/9/bits/hashtable_policy.h:2027:13:   required from ‘struct std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<const unsigned int, LootTemplate>, false> > >’
/usr/include/c++/9/bits/hashtable.h:173:11:   required from ‘class std::_Hashtable<unsigned int, std::pair<const unsigned int, LootTemplate>, std::allocator<std::pair<const unsigned int, LootTemplate> >, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >’
/usr/include/c++/9/bits/unordered_map.h:105:18:   required from ‘class std::unordered_map<unsigned int, LootTemplate>’
/home/mangos/cmangos/wotlk/src/game/Loot/LootMgr.h:294:25:   required from here
/usr/include/c++/9/bits/stl_pair.h:215:11: error: ‘std::pair<_T1, _T2>::second’ has incomplete type
  215 |       _T2 second;                /// @c second is a copy of the second object
      |           ^~~~~~
In file included from /home/mangos/cmangos/wotlk/src/game/Entities/Player.h:42,
                 from /home/mangos/cmangos/wotlk/src/game/Grids/GridNotifiers.h:28,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/CreatureAI.cpp:20:
/home/mangos/cmangos/wotlk/src/game/Loot/LootMgr.h:35:7: note: forward declaration of ‘class LootTemplate’
   35 | class LootTemplate;
      |       ^~~~~~~~~~~~
```
```
In file included from /usr/include/c++/9/bits/stl_algobase.h:64,
                 from /usr/include/c++/9/bits/specfun.h:45,
                 from /usr/include/c++/9/cmath:1927,
                 from /home/mangos/cmangos/wotlk/src/shared/Common.h:28,
                 from /home/mangos/cmangos/wotlk/src/framework/Dynamic/ObjectRegistry.h:22,
                 from /home/mangos/cmangos/wotlk/src/framework/Dynamic/FactoryHolder.h:24,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/UnitAI.h:23,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/CreatureAI.h:22,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/GuardAI.h:22,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/GuardAI.cpp:19:
/usr/include/c++/9/bits/stl_pair.h: In instantiation of ‘struct std::pair<const unsigned int, LootTemplate>’:
/usr/include/c++/9/ext/aligned_buffer.h:91:28:   required from ‘struct __gnu_cxx::__aligned_buffer<std::pair<const unsigned int, LootTemplate> >’
/usr/include/c++/9/bits/hashtable_policy.h:233:43:   required from ‘struct std::__detail::_Hash_node_value_base<std::pair<const unsigned int, LootTemplate> >’
/usr/include/c++/9/bits/hashtable_policy.h:279:12:   required from ‘struct std::__detail::_Hash_node<std::pair<const unsigned int, LootTemplate>, false>’
/usr/include/c++/9/bits/hashtable_policy.h:2027:13:   required from ‘struct std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<const unsigned int, LootTemplate>, false> > >’
/usr/include/c++/9/bits/hashtable.h:173:11:   required from ‘class std::_Hashtable<unsigned int, std::pair<const unsigned int, LootTemplate>, std::allocator<std::pair<const unsigned int, LootTemplate> >, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >’
/usr/include/c++/9/bits/unordered_map.h:105:18:   required from ‘class std::unordered_map<unsigned int, LootTemplate>’
/home/mangos/cmangos/wotlk/src/game/Loot/LootMgr.h:294:25:   required from here
/usr/include/c++/9/bits/stl_pair.h:215:11: error: ‘std::pair<_T1, _T2>::second’ has incomplete type
  215 |       _T2 second;                /// @c second is a copy of the second object
      |           ^~~~~~
In file included from /home/mangos/cmangos/wotlk/src/game/Entities/Player.h:42,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/GuardAI.cpp:21:
/home/mangos/cmangos/wotlk/src/game/Loot/LootMgr.h:35:7: note: forward declaration of ‘class LootTemplate’
   35 | class LootTemplate;
      |       ^~~~~~~~~~~~
In file included from /usr/include/c++/9/bits/stl_algobase.h:64,
                 from /usr/include/c++/9/bits/specfun.h:45,
                 from /usr/include/c++/9/cmath:1927,
                 from /home/mangos/cmangos/wotlk/src/shared/Common.h:28,
                 from /home/mangos/cmangos/wotlk/src/framework/Dynamic/ObjectRegistry.h:22,
                 from /home/mangos/cmangos/wotlk/src/framework/Dynamic/FactoryHolder.h:24,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/UnitAI.h:23,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/CreatureAI.h:22,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/PetAI.h:22,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/GuardianAI.h:22,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/GuardianAI.cpp:19:
/usr/include/c++/9/bits/stl_pair.h: In instantiation of ‘struct std::pair<const unsigned int, LootTemplate>’:
/usr/include/c++/9/ext/aligned_buffer.h:91:28:   required from ‘struct __gnu_cxx::__aligned_buffer<std::pair<const unsigned int, LootTemplate> >’
/usr/include/c++/9/bits/hashtable_policy.h:233:43:   required from ‘struct std::__detail::_Hash_node_value_base<std::pair<const unsigned int, LootTemplate> >’
/usr/include/c++/9/bits/hashtable_policy.h:279:12:   required from ‘struct std::__detail::_Hash_node<std::pair<const unsigned int, LootTemplate>, false>’
/usr/include/c++/9/bits/hashtable_policy.h:2027:13:   required from ‘struct std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<const unsigned int, LootTemplate>, false> > >’
/usr/include/c++/9/bits/hashtable.h:173:11:   required from ‘class std::_Hashtable<unsigned int, std::pair<const unsigned int, LootTemplate>, std::allocator<std::pair<const unsigned int, LootTemplate> >, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >’
/usr/include/c++/9/bits/unordered_map.h:105:18:   required from ‘class std::unordered_map<unsigned int, LootTemplate>’
/home/mangos/cmangos/wotlk/src/game/Loot/LootMgr.h:294:25:   required from here
/usr/include/c++/9/bits/stl_pair.h:215:11: error: ‘std::pair<_T1, _T2>::second’ has incomplete type
  215 |       _T2 second;                /// @c second is a copy of the second object
      |           ^~~~~~
In file included from /home/mangos/cmangos/wotlk/src/game/Entities/Player.h:42,
                 from /home/mangos/cmangos/wotlk/src/game/Grids/GridNotifiers.h:28,
                 from /home/mangos/cmangos/wotlk/src/game/AI/BaseAI/GuardianAI.cpp:22:
/home/mangos/cmangos/wotlk/src/game/Loot/LootMgr.h:35:7: note: forward declaration of ‘class LootTemplate’
   35 | class LootTemplate;
      |       ^~~~~~~~~~~~
```
```
make[2]: *** [src/game/CMakeFiles/game.dir/build.make:102: src/game/CMakeFiles/game.dir/AI/BaseAI/GuardianAI.cpp.o] Error 1
make[2]: *** [src/game/CMakeFiles/game.dir/build.make:128: src/game/CMakeFiles/game.dir/AI/BaseAI/PetAI.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:748: src/game/CMakeFiles/game.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 18%] Linking CXX static library libmmaplib.a
[ 18%] Built target mmaplib
make: *** [Makefile:130: all] Error 2
```

### Steps to Reproduce

1.Compile current 
2.
3.
4.

### Core SHA1 Commit Hash

CMaNGOS-Core revision : eff384a85689141e62bf7c58cbb7ad4e6c6738c7

### Database SHA1 Commit Hash

N/a

### Operating System

Ubuntu Server LTS 20.04

### Client Version

3.3.5a (Wrath of the Lich King)

**Comment by kelbren** [comment]
I was able to work around this by installing compiler version 13. Ubuntu is still on v9 in standard repos.
by using the instructions here: https://www.linuxcapable.com/how-to-install-gcc-compiler-on-ubuntu-linux/

hope this helps.

**Comment by evil-at-wow** [comment]
Using GCC 13 is really a workaround and other compilers might still give errors, and rightfully so. The code should be fixed: class ```LootStore``` has a member ```LootTemplateMap m_LootTemplates;``` (line 294 mentioned in the error) which is actually ```std::unordered_map<uint32, LootTemplate> LootTemplateMap;``` but the problem here is that ```LootTemplate``` at that point is still an incomplete type (there is only a forward declaration on line 35, and the class is declared later in the same header). Since C++17 it is allowed to instantiate ```std::vector``` and ```std::list``` with an incomplete type, but here we have ```std::unordered_map``` as a container. So your compiler rightfully refused to compile this.

I haven't checked their API in detail yet, but I think you can swap places for ```class LootStore``` and ```class LootTemplate``` in LootMgr.h to fix this.

[edit] I figured we should have seen an error in our CI builds, but they all passed apparently. I'm actually quite surprised all of the compilers on our CI accepted this!

**Comment by adellaci** [comment]
I did get it to compile with gcc/g++ (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0, but it is riddled with warnings

**Comment by adellaci** [comment]
I was originally started with Ubuntu 20.04 Server w/ gcc/g++ ver 9, I upgraded the server to Ubuntu 22.04 Server and have been working my way up GNU compiler version. 9 & 10 wouldn't not compile.  I skipped 11. 12 Compiled with 12 warnings and a couple 'deprecated' statements, mainly around 'Boost'. I may install ver 13, but under Ubuntu 22.04 it looks like it still under testing.

**Comment by adellaci** [comment]
@evil-at-wow what would end up being the preferred compilers, let me rephrase that, the targeted compilers

**Comment by evil-at-wow** [comment]
You certainly don't need the very latest compilers to get the cores to build. We're using C++17, so that means that the vast majority of the features is supported by GCC 7 or later (although if I remember correctly this is missing std::filesystem, so maybe I should say GCC 8), clang 4 or later, and the most recent Visual Studio 2017 (15.8). Those are minimal versions though, and using a more recent version is probably a good idea if available (perhaps with the exception of GCC 11, which is known to have some issue when compiling our cores in release mode). So Ubuntu 20.04 with GCC 9.x should be fine really, and so is 22.04 with either GCC 9 or 10.

Sometimes you'll get compile errors no matter what compiler you're using because someone actually broke the code (like in this case here). Similarly you'll most likely get some compiler warnings as well. There's quite a few remaining, depending on options, and new versions of compilers also introduce new ones all the time so it's a never ending battle to say so. I've been slowly working my way through, but in the meantime they are mostly that for users: warnings. I mean, there's no point in me quickly silencing a tons of warning when that introduces all kinds of bugs and regressions.

That said, I've noticed our CI has a tendency to use very recent versions so we're not exactly testing minimal or even more moderate requirements, which I think is quite unfortunate (yet understandable). As demonstrated in this issue that may lead to things building on CI with the lastest version but not some older versions that we actually still support.

**Comment by cyberium** [comment]
Should be fixed in https://github.com/cmangos/mangos-wotlk/commit/f156451702b37dcc178aa8732215c4cafacf6082
